### PR TITLE
hpa v2beta1 api warning on DC & edit autoscaler pages

### PR DESCRIPTION
**Pull request by benjaminapetersen**
This PR depends on [common 285](https://github.com/openshift/origin-web-common/pull/285).

[WIP] notes:

`bower link origin-web-common` with the above PR or the tests will fail. 

On the DC page:

screen shot 2018-01-23 at 6 21 58 pm

- probably only want to show the v2beta1 message, will hide the other

On the edit autoscaler page:

screen shot 2018-01-23 at 6 25 12 pm

- shows a message w/a link to CLI
- hides the CPU request input as editing this is probably not what the user wants to do

TODO (perhaps follow-on PR):

- Update the warning message to show the user the data from the `metrics` annotation so they can easily see what the autoscaler is using to scale the resource.

### CODE REVIEW

**Comment by benjaminapetersen** [comment]
Updated to drop the cpuRequest warning if already shows the v2HPA warning:

screen shot 2018-01-23 at 9 28 34 pm

screen shot 2018-01-23 at 9 28 50 pm

Thinking about best wording for these messages.  Also looking to see if we can reduce the duplicate templating.  The browse `deployment-config.html`, `deployment.html` & `_replica-set-details.html` look like they have a good bit of duplicated html around the hpa warnings, and the `oscAutoscaler` directive has similar but slightly different set for the same warnings.

**Comment by benjaminapetersen** [comment]
Tests will fail as I have not yet updated the ones I just added in that last PR.

**Comment by spadgett** [comment]
I'm not convinced this should be a warning on the browse pages since there's not really anything wrong. It should be a warning in the editor since it doesn't support the new fields, though.

@jwforres opinion?

**Comment by jwforres** [comment]
yeah warning should only have been in the editor

**Comment by benjaminapetersen** [comment]
Makes sense to me. Let me tweak.

File: app/scripts/controllers/deploymentConfig.js [hunk 1]
```diff
@@ -74,7 +74,8 @@ angular.module('openshiftConsole')
         var updateHPAWarnings = function() {
             HPAService.getHPAWarnings($scope.deploymentConfig, $scope.autoscalers, limitRanges, project)
                       .then(function(warnings) {
-              $scope.hpaWarnings = warnings;
+              // TODO: this should probably be a $filter we can reuse in other controllers?
```
Comment (benjaminapetersen) [inline app/scripts/controllers/deploymentConfig.js]: So I can pretty easily eliminate the warning from the controller, but thinking this should be a filter to share with other controllers.

File: app/scripts/controllers/deployment.js [hunk 2]
```diff
@@ -68,7 +68,7 @@ angular.module('openshiftConsole')
         var updateHPAWarnings = function() {
             HPAService.getHPAWarnings($scope.deployment, $scope.autoscalers, limitRanges, project)
```
Comment (benjaminapetersen) [inline app/scripts/controllers/deployment.js]: I added this filter to each of the deployment pages, but now think I probably should have just added a `HPAService.getHPAWarningsForResource()` method & eliminated the duplication.

File: app/scripts/filters/resources.js [hunk 3]
```diff
@@ -1354,4 +1354,11 @@ angular.module('openshiftConsole')
   })
   .filter('humanizePodStatus', function(humanizeReasonFilter) {
     return humanizeReasonFilter;
+  })
+  // not all of the HPA warnings should be surfaced on resources pages,
```
Comment (benjaminapetersen) [inline app/scripts/filters/resources.js]: Ben, you gonna finish the comment? or nah?

**Comment by benjaminapetersen** [comment]
Test updated, but im `bower link origin-web-common` locally, so the build will not pass.

File: test/spec/services/hpaServiceSpec.js [hunk 4]
```diff
@@ -22,8 +21,7 @@ describe('HPAService', function() {
   }));
 
 
-  beforeEach(inject(function(_$q_, _$rootScope_, _$window_, _HPAService_) {
-    $rootScope = _$rootScope_;
+  beforeEach(inject(function(_$q_, _$window_, _HPAService_) {
```
Comment (benjaminapetersen) [inline test/spec/services/hpaServiceSpec.js]: @spadgett also did remove the `$rootScope` `$digest`.  Thought that was necessary to get `$q` to resolve in this env, but guess not!

File: app/scripts/services/hpa.js [hunk 5]
```diff
@@ -146,15 +168,27 @@ angular.module("openshiftConsole")
         return $q.when([]);
       }
       return MetricsService.isAvailable().then(function(metricsAvailable) {
+        var v2HPAWarning = hasV2HPAScaler(scaleTarget, hpaResources);
         return _.compact([
           hasMetricsAvailableWarning(metricsAvailable),
-          hasCPURequestWarning(scaleTarget, limitRanges, project),
+          v2HPAWarning ? v2HPAWarning : hasCPURequestWarning(scaleTarget, limitRanges, project),
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: I think this would be easier to read as...

```js
hasMetricsAvailableWarning(metricsAvailable),
// Only check for CPU request if not using v2 HPA metrics.
hasV2HPAScaler(scaleTarget, hpaResources) || hasCPURequestWarning(scaleTarget, limitRanges, project),
hasCompetingAutoscalersWarning(hpaResources),
// ...
```

File: app/scripts/services/hpa.js [hunk 6]
```diff
@@ -99,6 +100,27 @@ angular.module("openshiftConsole")
...
 
 
+    var hasV2HPAAnnotations = function(hpaResources) {
+      return _.some(hpaResources, function(hpa) {
+        return $filter('hpaMetrics')(hpa);
+      });
+    };
+
+
+    var hasV2HPAScaler = function(scaleTarget, hpaResources) {
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: hasV2HPAScalerWarning to be consistent with the other functions?

File: app/scripts/services/hpa.js [hunk 7]
```diff
@@ -99,6 +100,27 @@ angular.module("openshiftConsole")
     };
 
 
+    var hasV2HPAAnnotations = function(hpaResources) {
+      return _.some(hpaResources, function(hpa) {
+        return $filter('hpaMetrics')(hpa);
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: Since this is temporary and we shouldn't need it after we update to the newer API, I'd just inline the full annotation here.

```js
return $filter('annotation')(hpa, 'autoscaling.alpha.kubernetes.io/metrics');
```

I'd rather not do more than we need to with the annotations if the same values are supported in the v2beta1 API. And then you don't need the common PR.

nit: better to resolve the filter outside the loop

File: app/scripts/filters/resources.js [hunk 8]
```diff
@@ -1355,4 +1355,9 @@ angular.module('openshiftConsole')
   })
   .filter('humanizePodStatus', function(humanizeReasonFilter) {
     return humanizeReasonFilter;
+  })
+  .filter('hpaWarningsForResource', function() {
```
Comment (spadgett) [inline app/scripts/filters/resources.js]: This is a very specific filter. I'd give it a name that makes it clear what it's really doing like `rejectV2HPAWarnings`. But we might not need it. (See my comments below.)

File: app/scripts/services/hpa.js [hunk 9]
```diff
@@ -146,15 +168,27 @@ angular.module("openshiftConsole")
...
     };
 
+    // Get HPA warnings relevant to a Resource list page.
+    //
+    // This method returns the same set as the above getHPAWarnings but
+    // filters out the warnings that don't need to be fully surfaced to
+    // each individual resource.
+    var getHPAWarningsForResource = function(scaleTarget, hpaResources, limitRanges, project) {
+      return getHPAWarnings(scaleTarget, hpaResources, limitRanges, project).then(function(warnings) {
+        return $filter('hpaWarningsForResource')(warnings);
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: nit: better to resolve the filter outside the loop

File: app/scripts/services/hpa.js [hunk 10]
```diff
@@ -185,10 +219,17 @@ angular.module("openshiftConsole")
       return hpaByResource;
     };
 
+    // currently unsupported by the webconsole
+    var isUnsupportedAPI = function(hpa) {
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: nit: `isUnsupportedAPI` sounds like the API is no longer supported by the server, but that's not the case. I think something like `usesV2Metrics` is more clear? The resource is technically v1 API since that's what we requested.

File: app/scripts/services/hpa.js [hunk 11]
```diff
@@ -146,15 +168,27 @@ angular.module("openshiftConsole")
...
         ]);
       });
     };
 
+    // Get HPA warnings relevant to a Resource list page.
+    //
+    // This method returns the same set as the above getHPAWarnings but
+    // filters out the warnings that don't need to be fully surfaced to
+    // each individual resource.
+    var getHPAWarningsForResource = function(scaleTarget, hpaResources, limitRanges, project) {
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: Are we filtering out the v2 warnings from all places we show them? For instance, the deployment donut on the overview?

I think it would be safer to opt in to the v2 editor warnings instead of always including editor warnings and having to filter them out. There's only one context where we need the extra warning: the editor. And that context performs the check in a different way (`isUnsupportedAPI()`).

File: app/views/directives/osc-autoscaling.html [hunk 12]
```diff
@@ -99,7 +99,7 @@
       </span>
     </div>
   </div>
-  <div class="form-group">
+  <div ng-show="showCPURequestInput"  class="form-group">
```
Comment (spadgett) [inline app/views/directives/osc-autoscaling.html]: nit: extra space before class

File: app/scripts/services/hpa.js (hunk 6 above)
Comment (spadgett) [reply app/scripts/services/hpa.js]: Unless I'm misunderstanding, we will be filtering this out every time we add it. The only place we want to show the warning is the editor page, but we specifically check `isUnsupportedAPI` there instead of calling the get warnings method. So I think we just avoid adding the warning in `getHPAWarnings`.

File: app/views/edit/autoscaler.html [hunk 13]
```diff
@@ -38,11 +38,21 @@ <h1>
...
+            </div>
+
+
             <fieldset ng-disabled="disableInputs" class="gutter-top">
               <osc-autoscaling
                   model="autoscaling"
                   show-name-input="true"
-                  name-read-only="kind === 'HorizontalPodAutoscaler'">
+                  name-read-only="kind === 'HorizontalPodAutoscaler'"
+                  show-cpu-request-input="showV2Beta1Warning">
```
Comment (spadgett) [inline app/views/edit/autoscaler.html]: nit: I'd call this scope var `usesV2Metrics` since we use the flag for more than just showing the warning.

**Comment by benjaminapetersen** [comment]
@spadgett 

    Just confirming that not all HPAs have the metrics annotation we're checking, only ones that scale on metrics other than CPU request.

I've been going off of the docs page @ `kubernetes.io`, but not surprised if I missed something.  Have a link handy?

File: app/scripts/filters/resources.js (hunk 8 above)
Comment (benjaminapetersen) [reply app/scripts/filters/resources.js]: Ok I can remove.

File: app/scripts/services/hpa.js (hunk 7 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: This makes sense as well, closed the other PR.

File: app/scripts/services/hpa.js (hunk 6 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: I can eliminate `hasV2HPAScalerWarning` from `getHPAWarnings`, though then I'll be back to showing this:

screen shot 2018-02-01 at 3 17 38 pm

Which I'll still need to check for `v2beta1` hpa annotations & filter out anyway.  I'm not sure if I can entirely remove the filtering problem.

File: app/scripts/services/hpa.js (hunk 11 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: Ah, I didn't check the donut, thanks, this is problematic:

screen shot 2018-02-01 at 3 22 10 pm
screen shot 2018-02-01 at 3 22 32 pm

I think I may still need to do the `hasV2HPAScalerWarning` check in `getHPAWarnings`, but then just drop both `v2` && the `CPURequest` warning.

File: app/scripts/services/hpa.js (hunk 9 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: I'll drop the `filter` per your other comment.

File: app/scripts/controllers/edit/autoscaler.js [hunk 14]
```diff
@@ -78,6 +78,21 @@ angular.module('openshiftConsole')
     var horizontalPodAutoscalerVersion = APIService.getPreferredVersion('horizontalpodautoscalers');
     var limitRangesVersion = APIService.getPreferredVersion('limitranges');
 
+    var getHPAWarnings = function(resource, limitRanges, project, context) {
+      var scaleTarget = resource.spec.scaleTargetRef;
+      var scaleTargetVersion = APIService.objectToResourceGroupVersion(scaleTarget);
+      // if the DC still has extensions in the verison, it will fail, so we need to
+      // get the preferred version.
```
Comment (benjaminapetersen) [inline app/scripts/controllers/edit/autoscaler.js]: Wish we had es6 fat arrows && auto `return` 😄

Comment (benjaminapetersen) [reply app/scripts/controllers/edit/autoscaler.js]: This gets the `edit` `autoscaler` page able to use `warnings` from the `hpaService` instead of the standalone `showCPURequestWarning` flags.

File: app/scripts/controllers/edit/autoscaler.js [hunk 15]
```diff
@@ -78,6 +78,21 @@ angular.module('openshiftConsole')
...
+      var scaleTargetPreferredVersion = APIService.getPreferredVersion(scaleTargetVersion.resource);
+      return DataService.get(scaleTargetPreferredVersion, scaleTarget.name, context).then(function(target) {
+        return HPAService.getHPAEditorWarnings(target, [resource], limitRanges, project).then(function(warnings) {
+          return _.keyBy(warnings, function(warning) {
```
Comment (spadgett) [inline app/scripts/controllers/edit/autoscaler.js]:
```js
return _.keyBy(warnings, 'reason');
```

File: app/scripts/controllers/edit/autoscaler.js [hunk 16]
```diff
@@ -78,6 +78,21 @@ angular.module('openshiftConsole')
     var horizontalPodAutoscalerVersion = APIService.getPreferredVersion('horizontalpodautoscalers');
     var limitRangesVersion = APIService.getPreferredVersion('limitranges');
 
+    var getHPAWarnings = function(resource, limitRanges, project, context) {
+      var scaleTarget = resource.spec.scaleTargetRef;
+      var scaleTargetVersion = APIService.objectToResourceGroupVersion(scaleTarget);
+      // if the DC still has extensions in the verison, it will fail, so we need to
+      // get the preferred version.
+      var scaleTargetPreferredVersion = APIService.getPreferredVersion(scaleTargetVersion.resource);
+      return DataService.get(scaleTargetPreferredVersion, scaleTarget.name, context).then(function(target) {
```
Comment (spadgett) [inline app/scripts/controllers/edit/autoscaler.js]: I don't think we can always get the scale target because it might be a scale subresource that doesn't support GET?

File: app/scripts/directives/oscFileInput.js [hunk 17]
```diff
@@ -127,6 +127,10 @@ angular.module('openshiftConsole')
             if (files.length > 0 ) {
               scope.file = _.head(files);
               addFile(scope.file);
+            } else {
+              scope.uploadError = true;
+              scope.cleanInputValues();
+              Logger.error("Could not read file", e);
```
Comment (spadgett) [inline app/scripts/directives/oscFileInput.js]: This change looks unrelated

File: app/views/edit/autoscaler.html [hunk 18]
```diff
@@ -38,11 +38,21 @@ <h1>
...
+            </div>
+
+
             <fieldset ng-disabled="disableInputs" class="gutter-top">
               <osc-autoscaling
                   model="autoscaling"
                   show-name-input="true"
-                  name-read-only="kind === 'HorizontalPodAutoscaler'">
+                  name-read-only="kind === 'HorizontalPodAutoscaler'"
+                  show-request-input="!(usesV2Metrics)">
```
Comment (spadgett) [inline app/views/edit/autoscaler.html]: I don't see where `usesV2Metrics` is set.

File: app/scripts/controllers/edit/autoscaler.js (hunk 16 above)
Comment (benjaminapetersen) [reply app/scripts/controllers/edit/autoscaler.js]: Is there a good example of that?   I'm thinking about things we have like `deploymentconfig/instantiate` or `pods/log` or `builds/clone` and can't think of a why an autoscaler would be attached.

File: app/scripts/directives/oscFileInput.js (hunk 17 above)
Comment (benjaminapetersen) [reply app/scripts/directives/oscFileInput.js]: Correct, cherry picked jakub's PR for the file input, should have saw before my commit, removing.

File: app/scripts/controllers/edit/autoscaler.js [hunk 19]
```diff
@@ -78,6 +78,7 @@ angular.module('openshiftConsole')
     var horizontalPodAutoscalerVersion = APIService.getPreferredVersion('horizontalpodautoscalers');
     var limitRangesVersion = APIService.getPreferredVersion('limitranges');
 
+
```
Comment (benjaminapetersen) [inline app/scripts/controllers/edit/autoscaler.js]: Rolled this back to minmal changes.

Note that  as it stands now & like `master`, the `showCPURequestWarning` will not trigger w/o hitting a certain set of conditionals, though the cpu request warning DOES appear on the DC page.

File: app/views/edit/autoscaler.html (hunk 18 above)
Comment (benjaminapetersen) [reply app/views/edit/autoscaler.html]: fixed

**Comment by benjaminapetersen** [comment]
@spadgett rolled back, looking over my tests again now, think just some minimal tweaks there.

**Comment by benjaminapetersen** [comment]
Ah, `$rootScope.$digest()` had to be introduced again, otherwise none of the test updates actually ran :frown:

File: app/scripts/services/hpa.js [hunk 20]
```diff
@@ -99,6 +102,27 @@ angular.module("openshiftConsole")
...
+    var hasV2HPAAnnotations = function(hpaResources) {
+      return _.some(hpaResources, function(hpa) {
+        return JSON.parse(annotation(hpa, 'autoscaling.alpha.kubernetes.io/metrics'));
+      });
+    };
+
+
+    var hasV2HPAScalerWarning = function(scaleTarget, hpaResources) {
+      var containers = _.get(scaleTarget, 'spec.template.spec.containers', []);
+      var hasRequest = hasCPURequest(containers);
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: It's not clear to how this affects whether we show the warning. Even if the container has a request, wouldn't we still show the warning if it's scaling on another metric we don't support in the editor?

File: app/scripts/services/hpa.js [hunk 21]
```diff
@@ -146,15 +170,19 @@ angular.module("openshiftConsole")
         return $q.when([]);
       }
       return MetricsService.isAvailable().then(function(metricsAvailable) {
+        var isV2HPA = hasV2HPAScalerWarning(scaleTarget, hpaResources);
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: Why not just use `hasV2HPAAnnotations` here and remove `hasV2HPAScalerWarning`?

File: app/scripts/services/hpa.js (hunk 20 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: Well, via the [trello card description](https://trello.com/c/EuQHe0LT) I was under the impression that we are drawing a line between `cpuRequest` and `metrics` for how we show this warning.

Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: But, I belive the API does transfer `metrics[]` that are CPU related to a `cpuRequest` if requested via `v1` api, so your comment makes sense.

File: app/scripts/services/hpa.js (hunk 21 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: Yeah that makes sense.

File: app/scripts/services/hpa.js (hunk 20 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: Eliminating this `fn`, then, as we don't need it at all, the `autoscaler` editor does it inline.

Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: On that note then, in the editor, we could keep showing the cpu input box for `targetCPUUtilizationPercentage` if that value is set, even if there are other metrics.

**Comment by benjaminapetersen** [comment]
So, editor at present-

With ONLY v2 metrics fields:

screen shot 2018-02-07 at 10 26 53 am

With v2 metrics && also a v2 metric on cpu that is converted back into a v1 `targetCPUUtilizationPercentage`:

screen shot 2018-02-07 at 10 38 33 am

With only the `targetCPUUtilizationPercentage` (just v1 fields):

screen shot 2018-02-07 at 10 39 50 am

Just hiding the input if there are ONLY v2 metrics present, otherwise showing it. Editing seems to work fine.  Viewing the objects under `other resources` & editing yaml appears correct with all 3 of these options.

**Comment by benjaminapetersen** [comment]
& Tests pass again.

File: app/scripts/services/hpa.js [hunk 22]
```diff
@@ -99,6 +102,13 @@ angular.module("openshiftConsole")
     };
 
 
+    var hasV2HPAAnnotations = function(hpaResources) {
+      return _.some(hpaResources, function(hpa) {
+        return JSON.parse(annotation(hpa, 'autoscaling.alpha.kubernetes.io/metrics'));
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: Missed this before (sorry), but if the annotation is not valid JSON -- for instance, if edited by the user -- it will be a runtime error. We should add a try/catch here if we need JSON, but I would just skip the parse since we don't do anything with the value.

File: app/scripts/services/hpa.js [hunk 23]
```diff
@@ -146,15 +156,19 @@ angular.module("openshiftConsole")
         return $q.when([]);
       }
       return MetricsService.isAvailable().then(function(metricsAvailable) {
+        var isV2HPA = hasV2HPAAnnotations(hpaResources);
```
Comment (spadgett) [inline app/scripts/services/hpa.js]: So looking at this again, it will skip CPU request validation if any of the resources uses v2 metrics. So if you have two:

1. v2 HPA
2. v1 HPA whose target has no CPU request

The CPU request warning will be dropped. Maybe that's OK since you really shouldn't have two HPA for the same resource, but pointing it out.

File: app/scripts/services/hpa.js (hunk 22 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: Hmm. will the save be successful if the user edits but the JSON is invalid?  Prob my bad assuming we would reject the change...

Comment (spadgett) [reply app/scripts/services/hpa.js]: Either way we don't want the runtime error in the client that stops JS execution.

File: app/scripts/services/hpa.js (hunk 23 above)
Comment (benjaminapetersen) [reply app/scripts/services/hpa.js]: Noting our discussion in case I come back to read these comments later, having 2 HPA is a problem that is more severe && MUST be fixed.  The CPU request warning might be a problem but is subordinate.

File: app/scripts/services/hpa.js [hunk 24]
```diff
@@ -99,6 +102,13 @@ angular.module("openshiftConsole")
     };
 
 
+    var hasV2HPAAnnotations = function(hpaResources) {
+      return _.some(hpaResources, function(hpa) {
+        return annotation(hpa, 'autoscaling.alpha.kubernetes.io/metrics');
```
Comment (benjaminapetersen) [inline app/scripts/services/hpa.js]: Dropped the `JSON.parse`

**Comment by benjaminapetersen** [comment]
Updated the `JSON.parse`, might be good now. maybe :)
